[8.0] add account_invoice_product_supplier_price_update

chafique-delli opened the pull request:
This module allows in the supplier invoice, to automatically update all products whose unit price on the line is different from the supplier price.

*  **account_invoice_supplierinfo_update** : provides main features to update supplierinfo from invoice line information
* **account_invoice_supplierinfo_update_discount** : glue module to update discount field, if 'product_supplierinfo_discount' is installed.
* **account_invoice_supplierinfo_update_variant** : glue module to update variant supplierinfo, if 'product_variant_supplierinfo' is installed.
*  **account_invoice_supplierinfo_update_on_validate** : extra module to force user to update supplierinfo data when user validates the invoice.

chafique-delli committed:
[ADD] account_invoice_product_cost_price_update module

elicoidal commented on account_invoice_product_supplier_price_update/README.rst:
```diff
@@ -0,0 +1,50 @@
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/account-invoicing/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed `feedback
```
You can simplify this part with the newest template

elicoidal commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,27 @@
+
+{
+    'name': 'Account Invoice Product Supplier Price Update',
+    'summary': 'From the supplier invoice, to automatic update all products '
```
In the supplier invoice, automatically update all products 
whose unit price on the line is different from the supplier price

elicoidal commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,27 @@
+    'author': 'Akretion, Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'application': False,
+    'installable': True,
```
probably not necessary as this is the default one

chafique-delli commented:
@elicoidal , thank you for the code review, I fixed this.

eLBati commented on account_invoice_product_supplier_price_update/README.rst:
```diff
@@ -0,0 +1,46 @@
+
+This module allow in the supplier invoice, automatically update all products
+whose unit price on the line is different from the supplier price.
+It created a new supplier price if there is not or it updates the first.
```
`created` --> `creates`

chafique-delli commented:
@eLBati , thank you for the code review, I fixed this.

pedrobaeza commented on account_invoice_product_supplier_price_update/README.rst:
```diff
@@ -0,0 +1,46 @@
+Update Supplier Price of product from Supplier Invoice
+======================================================
+
+This module allow in the supplier invoice, automatically update all products
```
s/allow/allows

pedrobaeza replied:
s/, automatically/, to automatically

pedrobaeza commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,26 @@
+    'depends': [
+        'account',
+        'purchase',
+        'product_variant_supplierinfo',
```
This module and the following are not real dependencies. Just conveniences, but the module will work the same, except for that features.

chafique-delli replied:
@pedrobaeza ,  thank you for the code review.
Indeed, the dependencies on the modules 'purchase' and 'product_variant_supplierinfo' are not necessary.

pedrobaeza replied:
And you don't need also `product_supplierinfo_tree_price_info`

chafique-delli replied:
@pedrobaeza ,in my code, I use the field 'unit_price note' of 'product.supplierinfo' which comes from the module 'product_supplierinfo_tree_price_info'.

pedrobaeza replied:
But that module is only for displaying information in an human readable form, not for using for another purposes. I would say to not using anything from this module.

pedrobaeza commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,26 @@
+    'installable': True,
+    'depends': [
+        'account',
+        'purchase',
```
purchase is not needed for having product.supplierinfo. They are in the product module.

chafique-delli commented:
Travis test is red because the test of  'account_invoice_merge_purchase' module failed.

sebastienbeau commented on account_invoice_product_supplier_price_update/account_invoice.py:
```diff
@@ -0,0 +1,62 @@
+        for line in self.invoice_line:
+            current_price_unit = [(6, 0, [])]
+            suppinfo = self.env['product.supplierinfo'].search([
+                ('product_tmpl_id', '=', line.product_id.product_tmpl_id.id),
```
you must depend on this module to get the right supplier https://github.com/OCA/product-variant/pull/32

sebastienbeau commented:
@chafique-delli can you fis the way to get the right supplier and add a test with
- a specific supplier (a supplier info attach to one variant)
- a generic supplier (a supplier info attach to the template => all variant)

sebastienbeau commented on account_invoice_product_supplier_price_update/account_invoice.py:
```diff
@@ -0,0 +1,69 @@
+    def _check_product_supplierprice(self):
+        lines = []
+        for line in self.invoice_line:
+            current_price_unit = [(6, 0, [])]
```
The code can be more simple here.
- first the current_price_unit do not need the 6, 0 as you always use the [0][2] after so you can remove it.
- then you do not need all of the variable suppinfo...

`
something like that:

```
def _check_product_supplierprice(self):                         
lines = []                                                             
for line in self.invoice_line:                                         
    current_price_unit = []
    supplier = self.partner_id.commercial_partner_id or self.partner_id
suppinfo = False
    current_price_uni = []
    for seller in line.product_id.seller_ids:                          
        if supplier == seller.name: 
            suppinfo = seller
            current_price_unit = [pricelist.price for p in suppinfo.pricelist_ids]
            break                                                      
    if line.price_unit not in current_price_unit:                                               
        lines.append((0, 0, {                                          
            'name': line.name,                                         
            'current_price_unit': suppinfo.unit_price_note,            
            'new_price_unit': line.price_unit,                         
            'suppinfo_id': suppinfo.id,                                
            'to_variant': True,                                        
            'supplier_id': supplier_id.id,                             
            'product_id': line.product_id.id,                          
            'product_tmpl_id': line.product_id.product_tmpl_id.id,     
        }))                                                            
return lines                                                           
```

`

sebastienbeau commented on account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py:
```diff
@@ -0,0 +1,76 @@
+                self.env['pricelist.partnerinfo'].create(vals)
+
+
+class SupplierinvoiceUpdateProductSupplierpriceLine(models.TransientModel):
```
can you found a shorter class name?

sebastienbeau commented on account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py:
```diff
@@ -0,0 +1,76 @@
+                    vals.update({
+                        'product_tmpl_id': line.product_id.product_tmpl_id.id,
+                    })
+                line.suppinfo_id = self.env['product.supplierinfo'].create(
```
can you avoid the repetition of self.env['product.supplierinfo'] by creating a variable called supinfo_obj ?

sebastienbeau commented on account_invoice_product_supplier_price_update/wizard/supplier_invoice_update_product_supplier_price.py:
```diff
@@ -0,0 +1,76 @@
+import openerp.addons.decimal_precision as dp
+
+
+class SupplierinvoiceUpdateProductSupplierprice(models.TransientModel):
```
can you find a shorter class name?

sebastienbeau reviewed (changes requested):
Some code simplification should be done

sebastienbeau commented on account_invoice_product_supplier_price_update/tests/test_supplier_invoice_update_product_supplier_price.py:
```diff
@@ -0,0 +1,82 @@
+from openerp.tests.common import TransactionCase
+
+
+class TestSupplierInvoiceUpdateProductSupplierPrice(TransactionCase):
```
We should split the test in 3 as there is 3 test on 3 invoice

sebastienbeau reviewed (commented):
split test

sebastienbeau commented on account_invoice_product_supplier_price_update/tests/test_update_supplier_price.py:
```diff
@@ -0,0 +1,79 @@
+
+        # supplier invoice with pricelist supplierinfo to update and
+        # product supplierinfo is on product_product
+        invoice_7 = self.env.ref(
```
Can you split the test ? one per invoice

chafique-delli commented:
@gurneyalex, Is you can watch why the runbot test shows me this error:
except_orm: ('Error', u"You try to install module 'account_invoice_product_supplier_price_update' that depends on module 'product_variant_supplierinfo'.\nBut the latter module is not available in your system.")
While the file 'oca_dependencies.txt' contains' product-variant' repository which contains  product_variant_supplierinfo' module.
Thank you.

gurneyalex commented:
@chafique-delli runbot should be fixed now.

chafique-delli commented:
@gurneyalex , thank you for the fix on runbot, but is you can watch why the travis test shows an error on installing the  account_invoice_merge_purchase module whereas,  a priori my  account_invoice_product_supplier_price_update module is not installed yet.

gurneyalex commented:
@chafique-delli this is not your fault. The 8.0 branch build is red on Travis (https://travis-ci.org/OCA/account-invoicing/jobs/166920774)  -> #185

legalsylvain commented on account_invoice_product_supplier_price_update/wizard/update_supplier_price.py:
```diff
@@ -0,0 +1,81 @@
+    def update_product_supplierprice(self):
+        invoice = self.env['account.invoice'].browse(
+            self._context['active_id'])
+        invoice.signal_workflow('invoice_open')
```
Why do you launch this workflow ?

legalsylvain commented on account_invoice_product_supplier_price_update/wizard/update_supplier_price.py:
```diff
@@ -0,0 +1,81 @@
+
+    @api.multi
+    def update_product_supplierprice(self):
+        invoice = self.env['account.invoice'].browse(
```
self.ensure_one() is required if you do not manage a for each wizard in self.

legalsylvain commented on account_invoice_product_supplier_price_update/wizard/update_supplier_price.py:
```diff
@@ -0,0 +1,81 @@
+        invoice.signal_workflow('invoice_open')
+        for line in self.wizard_line_ids:
+            if not line.suppinfo_id:
+                vals = {
```
maybe calling default value function could allow to have different user defined default value ? (not sure about that point, just a question).

legalsylvain commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+    'depends': [
+        'account',
+        'product_supplierinfo_tree_price_info',
+        'product_variant_supplierinfo'
```
Hi. I understand why you depend on this module, but for me, the feature should work if product_variant_supplierinfo is not installed.
What about to split the module in two : 
-  account_invoice_product_supplier_price_update that will depends on account only and work with generic feature
- account_invoice_variant_product_supplier_price_update that will depends on 'product_variant_supplierinfo' and 'account_invoice_product_supplier_price_update', managing variants specifics, overloading function.

If people are agree with that point, I can help you for that refactoring.

legalsylvain commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+    'installable': True,
+    'depends': [
+        'account',
+        'product_supplierinfo_tree_price_info',
```
I don't see why you depend on this module. (even if this module is cool ;-) )

legalsylvain commented on account_invoice_product_supplier_price_update/__openerp__.py:
```diff
@@ -0,0 +1,29 @@
+    ],
+    'demo': [
+        'demo/product_demo.xml',
+        'demo/account_invoice_demo.xml',
```
could you git mv for
- demo/product_supplierinfo.xml
- demo/account_invoice.xml

[Ref](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#file-naming)

legalsylvain commented on account_invoice_product_supplier_price_update/account_invoice.py:
```diff
@@ -0,0 +1,60 @@
+            current_price_unit = []
+            suppinfo = self.env['product.supplierinfo'].browse(False)
+            supplier = self.partner_id.commercial_partner_id or self.partner_id
+            for seller in line.product_id.seller_ids:
```
something like that should more rapid:

```
for seller in line.product_id.seller_ids.filtered(lambda seller: seller.name== supplier):
```

legalsylvain commented on account_invoice_product_supplier_price_update/account_invoice.py:
```diff
@@ -0,0 +1,60 @@
+            for seller in line.product_id.seller_ids:
+                if supplier == seller.name:
+                    suppinfo = seller
+                    current_price_unit = [pricelist.price for pricelist in
```
Not a clear idea here, but I think we should think this algorithm as overloadable.

I think to [product_supplierinfo_discount](https://github.com/OCA/purchase-workflow/tree/10.0/product_supplierinfo_discount)

legalsylvain reviewed (changes requested):
Hi, @chafique-delli. Some remarks inline.  2 others below. Just a code review for the time being, I'll test it functionaly and more deeply friday.

Otherwise, thanks a lot to share this work !
For shops it is a "must have feature".
- Not sure the image module will help people to understand what the module makes. Default OCA module should be better ? (not a blocking point for me).
- Not sure to have understand all the workflow stuff : invoice validation... (but I just do a quick code review, I'll take a time friday).

legalsylvain commented:
Hi @chafique-delli,
I took a deeply look on your PR. Some remarks :
* This module should be renamed better in account_invoice_supplierinfo_update because it's not only the price that is updated. (min_qty should be too)
* I still think that it should not depend on variant settings because it adds an unncessary dependency. It should be managed by glue module. The module you're sharing is very generic and is a must have feature, and supplierinfo for variant is specific. The current design will avoid some guys to install this module, or to have to disable manually some undesired features.
* This module doesn't work in all situation. Specially if there are some supplierinfo defined. (min_qty is not managed correctly).

I'm working on a PR on your PR, refactoring your PR. 

Thanks.

legalsylvain committed:
[REF] full refactoring:
- OCA convention
- module renamed
- modularity in function. (prepare function added)
[ADD] description
[FIX] remove unused dependencies for the time being

legalsylvain commented:
hi all, 

FYI

If you're interested by this module, a pending refactoring is available here. 
https://github.com/akretion/account-invoicing/pull/3

moylop260 commented on oca_dependencies.txt:
```diff
@@ -2,3 +2,5 @@ account-financial-tools
 stock-logistics-workflow
 product-attribute
 bank-payment
+product-variant
```
Where are required these repositories?

legalsylvain replied:
Thanks for pointing this unused dependencies. I just fixed it.

legalsylvain commented:
Hi. 
I just rebased this PR, after the last merge. 
@chafique-delli, could you squach your commits ? (I can make it if you want).
@sebastienbeau, @all could you review this PR updated ?

Thanks !

chafique-delli commented:
@legalsylvain , I want you to do it.Thank you.

legalsylvain commented:
Note : Travis is red because upstream 8.0 branch is red.

chafique-delli committed:
add tests

chafique-delli committed:
add account_invoice_supplierinfo_update_variant module

chafique-delli committed:
fix README.rst file

legalsylvain committed:
Merge pull request #5 from akretion/8.0-supplierinfo-update-on-validate

add account_invoice_supplierinfo_on_validate module

chafique-delli committed:
fix openerp.py file

chafique-delli committed:
fix method to put value of to_variant field

legalsylvain committed:
Merge pull request #4 from akretion/8.0-supplierinfo-update-variant

add account_invoice_supplierinfo_update_variant module

chafique-delli committed:
add product-variant repository in oca_dependencies.txt

sebastienbeau commented:
Great work @legalsylvain @chafique-delli  ! Thx

legalsylvain commented:
This PR has been strongly refactored and is now validated by 2 people. ( @sebastienbeau and me), but as I wrote a lot of the refactored code, I'd prefer to have a full external :+1: 

@eLBati , @pedrobaeza , @elicoidal , @moylop260 : is someone interested to finish to review that PR that you began to review ?

kind regards.

moylop260 commented on account_invoice_supplierinfo_update/models/account_invoice_line.py:
```diff
@@ -0,0 +1,74 @@
+        self.ensure_one()
+        if not len(supplierinfo.pricelist_ids):
+            return False
+        else:
```
If you have a `if-return` then the `else` is not needed

moylop260 commented on account_invoice_supplierinfo_update/models/account_invoice_line.py:
```diff
@@ -0,0 +1,74 @@
+        """Given an invoice line, return the supplierinfo that matches
+        with product and supplier, if exist"""
+        self.ensure_one()
+        supplierinfos = (self.product_id.seller_ids.filtered(
```
Unnecessary parents `(`

moylop260 commented on account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py:
```diff
@@ -0,0 +1,78 @@
+    @api.depends('current_price', 'new_price')
+    @api.multi
+    def _compute_price_variation(self):
+        for line in self:
```
In order to reduce cyclomatic complexity (and quantity of tests to cover the code) you could remove `if-else` sentence using filtered and global write:
```python
self.write({'price_variation': False})
for line in self.filtered('current_price'):
     line.price_variation = 100 * (line.new_price - line.current_price) / line.current_price
```

moylop260 commented on account_invoice_supplierinfo_update_discount/models/account_invoice_line.py:
```diff
@@ -0,0 +1,24 @@
+    def _prepare_supplier_wizard_line(self, supplierinfo, partnerinfo):
+        res = super(AccountInvoiceLine, self)._prepare_supplier_wizard_line(
+            supplierinfo, partnerinfo)
+        res['current_discount'] = partnerinfo and partnerinfo.discount or False
```
new api supports BrowseNull.value <- `False`
Use: `res['current_discount'] = partnerinfo.discount`

legalsylvain replied:
Your proposal doesn't work.

![bool_error](https://cloud.githubusercontent.com/assets/3407482/21312356/d2d85164-c5eb-11e6-83d6-24ce0cc178ab.png)

But the following line does work.

        res['current_discount'] = partnerinfo and partnerinfo.discount

moylop260 commented on account_invoice_supplierinfo_update_variant/wizard/wizard_update_invoice_supplierinfo_line.py:
```diff
@@ -0,0 +1,26 @@
+    def _prepare_supplierinfo(self):
+        if self.to_variant:
+            return {
+                'product_id': self.product_id.id,
```
https://github.com/OCA/account-invoicing/blob/d0987a68bfc9c86a59e811c6262327d71cbafcb1/account_invoice_supplierinfo_update/wizard/wizard_update_invoice_supplierinfo_line.py#L62-L69

The original method have the same values but changing just `product_id`:
```python
    @api.multi
    def _prepare_supplierinfo(self):
        self.ensure_one()
        return {
            'product_tmpl_id': self.product_id.product_tmpl_id.id,
            'name': self.wizard_id.invoice_id.supplier_partner_id.id,
            'min_qty': 0.0,
            'delay': 1,
        }
```

What about use result from `super` and change to custom values required in pro of easiest inheriting?

`res = supers(...); res.update({...}); res.pop()`

(I don't know maybe is not possible)

moylop260 reviewed (approved):
👍 
Thanks!

Just small comments non-blocking

chafique-delli committed:
fix the inherited method _prepare_supplierinfo

legalsylvain committed:
[IMP] price_variation computation and some optimizations
[ADD] correct group in demo mode

legalsylvain commented:
@moylop260 : @chafique-delli and me updated our PR, according your last remarks.
Is it now all ok for you ? 

kind regards.

legalsylvain committed:
[FIX] set supplierinfo_ok to False, if data is copied

legalsylvain commented:
Ping @moylop260 !

(And happy New year !)

legalsylvain commented:
Previous review was OK. 3 :+1:. merging.
